Sources/Sentry/SentryNetworkTracker.m

```diff
@@ -265,6 +288,121 @@ - (void)urlSessionTask:(NSURLSessionTask *)sessionTask setState:(NSURLSessionTas
+        SENTRY_LOG_DEBUG(
+            @"captureFailedRequestsEnabled is disabled, not capturing HTTP Client errors.");
+        return;
+    }
+
+    // if request or response are null, we can't raise the event
+    if (sessionTask.currentRequest == nil || sessionTask.response == nil) {
+        SENTRY_LOG_DEBUG(@"Request or Response are null, not capturing HTTP Client errors.");
+        return;
+    }
+    // some properties are only available if the response is of the NSHTTPURLResponse type
+    // bail if not
+    if (![sessionTask.response isKindOfClass:[NSHTTPURLResponse class]]) {
+        SENTRY_LOG_DEBUG(@"Response isn't a known type, not capturing HTTP Client errors.");
+        return;
+    }
+    NSHTTPURLResponse *myResponse = (NSHTTPURLResponse *)sessionTask.response;
+    NSURLRequest *myRequest = sessionTask.currentRequest;
+    NSNumber *responseStatusCode = @(myResponse.statusCode);
+
+    if (![self containsStatusCode:myResponse.statusCode]) {
+        SENTRY_LOG_DEBUG(@"Response status code isn't within the allowed ranges, not capturing "
+                         @"HTTP Client errors.");
+        return;
+    }
+
+    if (![self isTargetMatch:myRequest.URL withTargets:SentrySDK.options.failedRequestTargets]) {
+        SENTRY_LOG_DEBUG(
+            @"Request url isn't within the request targets, not capturing HTTP Client errors.");
+        return;
+    }
+
+    NSString *message = [NSString
+        stringWithFormat:@"HTTP Client Error with status code: %ld", (long)myResponse.statusCode];
+
+    SentryEvent *event = [[SentryEvent alloc] initWithLevel:kSentryLevelError];
+
+    SentryThreadInspector *threadInspector = SentrySDK.currentHub.getClient.threadInspector;
+    NSArray<SentryThread *> *threads = [threadInspector getCurrentThreads];
+
+    // sessionTask.error isn't used because it's not about network errors but rather
+    // requests that are considered failed depending on the HTTP status code
+    SentryException *sentryException = [[SentryException alloc] initWithValue:message
+                                                                         type:@"HTTPClientError"];
+    sentryException.mechanism = [[SentryMechanism alloc] initWithType:@"HTTPClientError"];
+
+    for (SentryThread *thread in threads) {
+        if ([thread.current boolValue]) {
+            SentryStacktrace *sentryStacktrace = [thread stacktrace];
+            sentryStacktrace.snapshot = @(YES);
+
+            sentryException.stacktrace = sentryStacktrace;
+
+            break;
+        }
+    }
+
+    SentryRequest *request = [[SentryRequest alloc] init];
+
+    NSURL *url = [[sessionTask currentRequest] URL];
+
+    NSString *urlString = [NSString stringWithFormat:@"%@://%@%@", url.scheme, url.host, url.path];
+
+    request.url = urlString;
+    request.method = myRequest.HTTPMethod;
+    request.fragment = url.fragment;
+    request.queryString = url.query;
+    if (sessionTask.countOfBytesSent != 0) {
```

Comment:
`m`: Why don't we want to add this info when `countOfBytesSent` is `0`?

Reply:
0 means no body, GET has no body, and sending 0 is semantically wrong.

Reply:
This is the type of behaviour that should be delegated to 'SentryRequest'. 
Let's say we use SentryRequest in more then one place, everywhere we use it, we have to do this check. 
In other hand, if we only check this during SentryRequest serialization, we are safe no matter how the class is used.